medusa/storage/s3_compat.py

```diff
@@ -0,0 +1,202 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import base64
+import logging
+import os
+import io
+import itertools
+import subprocess
+import json
+from subprocess import PIPE
+from dateutil import parser
+from pathlib import Path
+
+from libcloud.storage.drivers.minio import MinIOStorageDriver
+
+from medusa.storage.abstract_storage import AbstractStorage
+import medusa.storage.s3_compat.concurrent
+from medusa.storage.s3_compat.awscli import AwsCli
+
+import medusa
+
+
+"""
+    S3BaseStorage supports all the S3 compatible storages. Certain providers might override this method
+    to implement their own specialities (such as environment variables when running in certain clouds)
+
+    This implementation uses awscli instead of libcloud's s3's driver for uploads/downloads. If you wish
+    to use the libcloud's internal driver instead of awscli dependency, select s3_rgw.
+"""
+
+
+class S3BaseStorage(AbstractStorage):
+
+    def connect_storage(self):
+        with io.open(os.path.expanduser(self.config.key_file), 'r', encoding='utf-8') as json_fi:
+            credentials = json.load(json_fi)
+
+        # MinIOStorageDriver is the only clean implementation of BaseS3StorageDriver in libcloud
+        driver = MinIOStorageDriver(
```

Comment:
> 
> 
> Apparently there was [an issue](https://github.com/apache/libcloud/issues/1539) with the region setting in the MinIOStorageDriver, which was fixed by creating [this custom class](https://github.com/thelastpickle/cassandra-medusa/pull/246/files#diff-c3d3aefd89ed037caf0585e9927e2628eab86f6e33e2ac61fb77683bd9d748ce).
> What's your take on this? Shouldn't we use the driver implemented in #246 for now, until this is clarified?
> @DrissiReda, I saw you closed [the libcloud issue](https://github.com/apache/libcloud/issues/1539). Is it because it was related to a configuration problem or was it because you were happy with the custom driver you created for MinIO in your medusa PR?

Custom driver wasn't enough, like I said in that issue, the libcloud driver overwrites the region as an empty string, so I had to modify that for it to work, even in libcloud it's not exposed to the configuration, so a hardcoded fix like what I did is needed for minIO, for other s3 providers, you can't do this, since the region can change, and you'll have to expose it as a parameter. 

This is the fix I'm referring to:
 Setting https://github.com/apache/libcloud/blob/4370a7021c11c1c61afe6470d2baafc093bfa0dd/libcloud/storage/drivers/minio.py#L49 
`region_name = "us-east-1"`

Reply:
I don't think we use the driver in cases where that happens, since the awscli is used for download / upload?

Reply:
At least I didn't hit it yet..